Push Metadata changes to a public repo

CodexRaunak opened the pull request:
#1046 
Current workings after the changes:
- Initialize git with the local plugin directory where all the metadata is stored locally.
- After running the recipe, collect the modernization metadata and push it. Currently for testing I am pushing it to https://github.com/CodexRaunak/metadata-plugin-modernizer-tool.
- Make a PR with title and body to the repo.

Current Problems:
- It's also committing and pushing empty changes.
- We are not forking the repo, rather just checking out branch and make a PR to the original.
I am working on these 2 issues.

Also when the metadata is pushed, it is not formated and looks like:
![image](https://github.com/user-attachments/assets/d6cbea46-68cf-44d2-b4c8-7ea8663c8e68)

After these changes now we have separate methods for interacting with git in `GHService` class, one to interact with the plugin and one to interact with the `metadata repo`. For example like `openPullRequest()` for plugin and `openMetadataPullRequest()` for `metadata repo`.
This may cause some duplicity, but we can see as we progress to keep it separate or use a single method for both by passing some params.
For now maybe keep it separate and see where this thing can go to make changes easily.

Any feedbacks or suggestions are welcomed
Thank You 

### Testing done
Able to open PR at https://github.com/CodexRaunak/metadata-plugin-modernizer-tool. 
Need to add test coverage in the tool.

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

CodexRaunak commented:
```
 java -jar ./plugin-modernizer-cli/target/jenkins-plugin-modernizer-999999-SNAPSHOT.jar dry-run --plugins tuleap-oauth -r UpgradeNextMajorParentVersion      
Run Plugin Modernizer in dry-run mode 
Fetching plugin code locally tuleap-oauth... 
locking FileBasedConfig[C:\Users\DELL PC\.config\jgit\config] failed after 5 retries 
Fetched repository from https://github.com/jenkinsci/tuleap-oauth-plugin.git to branch refs/heads/master 
Collecting metadata for plugin tuleap-oauth... Please be patient 
Done 
Running recipes io.jenkins.tools.pluginmodernizer.UpgradeNextMajorParentVersion for plugin tuleap-oauth... Please be patient 
Done 
Modernization metadata for plugin tuleap-oauth: C:\Users\DELL PC\.cache\jenkins-plugin-modernizer-cli\tuleap-oauth\modernization-metadata 
Successfully initialized git repository for plugin tuleap-oauth at C:\Users\DELL PC\.cache\jenkins-plugin-modernizer-cli\tuleap-oauth 
Set remote URL to https://github.com/CodexRaunak/metadata-plugin-modernizer-tool 
Added files to staging area: modernization-metadata 
Can only sign commit using SSH key. Skipping sign commit when using GH_TOKEN 
Changes committed successfully with message: Add Modernization metadata for plugin tuleap-oauth 
Pushed changes to repository for metadata on branch tuleap-oauth 
Pull request already exists: https://github.com/CodexRaunak/metadata-plugin-modernizer-tool/pull/6 
Pull request update: https://github.com/CodexRaunak/metadata-plugin-modernizer-tool/pull/6 
Using minimum JDK 11 from metadata for plugin tuleap-oauth
Build failed 
Skipping formatting for plugin tuleap-oauth as it is not using Spotless
Verifying plugin tuleap-oauth with JDK 11... Please be patient
Build failed 
Done
Plugin tuleap-oauth failed to verify with JDK 11
Plugin tuleap-oauth verified successfully with JDK 11
Collecting metadata for plugin tuleap-oauth... Please be patient 
Done 
Skipping commits changes for plugin tuleap-oauth in dry-run mode 
Skipping forking plugin tuleap-oauth in dry-run mode
Skipping sync plugin tuleap-oauth in dry-run mode
Skipping push changes for plugin tuleap-oauth in dry-run mode
Skipping pull request changes for plugin tuleap-oauth in dry-run mode
*************
Plugin: tuleap-oauth 
Dry run mode. Changes were made on C:\Users\DELL PC\.cache\jenkins-plugin-modernizer-cli\tuleap-oauth\sources but not committed
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/checks/IDTokenChecker.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/helper/TuleapAuthorizationCodeUrlBuilderImpl.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/helper/TuleapAuthorizationCodeUrlBuilder.java
Modified file: src/test/java/io/jenkins/plugins/tuleap_oauth/checks/IDTokenCheckerImplTest.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/checks/AuthorizationCodeChecker.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/checks/IDTokenCheckerImpl.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/checks/AuthorizationCodeCheckerImpl.java
Modified file: src/test/java/io/jenkins/plugins/tuleap_oauth/checks/AuthorizationCodeCheckerImplTest.java
Modified file: pom.xml
Modified file: src/test/java/io/jenkins/plugins/tuleap_oauth/helper/TuleapAuthorizationCodeUrlBuilderImplTest.java
Modified file: src/test/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealmTest.java
*************

```

There maybe so many logs right now about pushing the metadata, but for the time being we are printing all these at info level, later we can switch to debug.

gounthar commented:
```
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import java.nio.file.Files;
import java.nio.file.Paths;
[...]
public void prettyPrintJsonFile(String inputPath, String outputPath) throws IOException {
    ObjectMapper mapper = new ObjectMapper();
    Object json = mapper.readValue(Files.readAllBytes(Paths.get(inputPath)), Object.class);
    ObjectWriter writer = mapper.writerWithDefaultPrettyPrinter();
    writer.writeValue(Paths.get(outputPath).toFile(), json);
}
```

before sending the PR?

CodexRaunak commented:
After forking the metadata repo, we would be cloning it as well, so what would be the directory structure should look like, here's my thought:
```
.cache/
└── jenkins-plugin-modernizer-cli/
    ├── plugin1
    ├── plugin2
    ├── plugin3
    ├── plugin4
    ├── metadata-plugin-modernizer/
    │   ├── plugin1/
    │   │   └── modernization-metadata
    │   ├── plugin2/
    │   │   └── modernization-metadata
    │   ├── plugin3/
    │   │   └── modernization-metadata
    │   └── plugin4/
    │       └── modernization-metadata
    ├── health-score
    ├── plugin-installation-stats
    ├── plugins-versions
    └── update-center

```
Here the `metadata-plugin-modernizer`, we would be cloning it, a test repo for it -https://github.com/Raunak80Madan/metadata-plugin-modernizer.

What do you think, is it fine for now? Please do suggest any changes needed or if I missed anything.
Thank You

jonesbusy commented:
See https://github.com/jenkins-infra/plugin-modernizer-tool/blob/main/plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/utils/JsonUtils.java for JSON serialization (it used GSON internally, not Jackson). Feel free to add an option for pretty print output

For the moment I think we can keep the same structure of the cache. We just need to ensure there is not too much conflicts when we open PRs

CodexRaunak commented on plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/github/GHService.java:
```diff
@@ -971,6 +1089,56 @@ public void pushChanges(Plugin plugin) {
+     * @param plugin The plugin that have been modernized
+     */
+    public void pushMetadataChanges(Plugin plugin) {
+        //        if (config.isDryRun()) {
```
I was actually doing some experiments, I will perform a clean up

CodexRaunak commented:
> For the moment I think we can keep the same structure of the cache. We just need to ensure there is not too much conflicts when we open PRs

The structure would be same, just as we need to clone the metadata repo, I was trying to clone it at the root of the cache, rather than in the plugins directory, so would this be fine https://github.com/Raunak80Madan/metadata-plugin-modernizer?

CodexRaunak commented:
Changes made:

- Add syncing, forking, cloning, before making PR
- It's not commit or pushing any empty changes now.
- Removed comments, performed clean up
- To store the metadata uniquely for every time, I added a UTC current date and time to the name of the metadata file which will be pushed to the metadata repository. Need to think if and how to consolidate?

PRs being made at  https://github.com/Raunak80Madan/metadata-plugin-modernizer

Things still needed:
- Pretty printing
- Tests

LOGS:
```
 java -jar ./plugin-modernizer-cli/target/jenkins-plugin-modernizer-999999-SNAPSHOT.jar dry-run --plugins tuleap-oauth -r UpgradeNextMajorParentVersion      
Run Plugin Modernizer in dry-run mode 
Fetching plugin code locally tuleap-oauth... 
locking FileBasedConfig[C:\Users\DELL PC\.config\jgit\config] failed after 5 retries 
Collecting metadata for plugin tuleap-oauth... Please be patient 
Done 
Running recipes io.jenkins.tools.pluginmodernizer.UpgradeNextMajorParentVersion for plugin tuleap-oauth... Please be patient 
Done 
Modernization metadata for plugin tuleap-oauth: C:\Users\DELL PC\.cache\jenkins-plugin-modernizer-cli\tuleap-oauth\modernization-metadata 
Fetching metadata repo locally tuleap-oauth... 
Forking metadata of tuleap-oauth locally from repo metadata-plugin-modernizer... 
Synced the forked metadata repository for plugin tuleap-oauth 
Copied plugin tuleap-oauth modernization metadata to cache: C:\Users\DELL PC\.cache\jenkins-plugin-modernizer-cli\metadata-plugin-modernizer\tuleap-oauth\modernization-metadata-2025-06-09T01-19-50
Set remote URL to https://github.com/CodexRaunak/metadata-plugin-modernizer 
Added files to staging area: . 
Can only sign commit using SSH key. Skipping sign commit when using GH_TOKEN 
Changes committed successfully with message: Add Modernization metadata for plugin tuleap-oauth
Pushed changes to repository for metadata on branch tuleap-oauth
Pull request created: https://github.com/Raunak80Madan/metadata-plugin-modernizer/pull/2
Using minimum JDK 11 from metadata for plugin tuleap-oauth
Build failed
Skipping formatting for plugin tuleap-oauth as it is not using Spotless
Verifying plugin tuleap-oauth with JDK 11... Please be patient
Build failed
Done
Plugin tuleap-oauth failed to verify with JDK 11
Plugin tuleap-oauth verified successfully with JDK 11
Collecting metadata for plugin tuleap-oauth... Please be patient
Done 
Skipping commits changes for plugin tuleap-oauth in dry-run mode 
Skipping forking plugin tuleap-oauth in dry-run mode 
Skipping sync plugin tuleap-oauth in dry-run mode 
Skipping push changes for plugin tuleap-oauth in dry-run mode 
Skipping pull request changes for plugin tuleap-oauth in dry-run mode 
************* 
Plugin: tuleap-oauth 
Dry run mode. Changes were made on C:\Users\DELL PC\.cache\jenkins-plugin-modernizer-cli\tuleap-oauth\sources but not committed
Modified file: src/test/java/io/jenkins/plugins/tuleap_oauth/helper/TuleapAuthorizationCodeUrlBuilderImplTest.java
Modified file: pom.xml
Modified file: src/test/java/io/jenkins/plugins/tuleap_oauth/checks/AuthorizationCodeCheckerImplTest.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/checks/AuthorizationCodeCheckerImpl.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/checks/IDTokenCheckerImpl.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/checks/AuthorizationCodeChecker.java
Modified file: src/test/java/io/jenkins/plugins/tuleap_oauth/checks/IDTokenCheckerImplTest.java 
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/helper/TuleapAuthorizationCodeUrlBuilder.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealm.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/helper/TuleapAuthorizationCodeUrlBuilderImpl.java
Modified file: src/main/java/io/jenkins/plugins/tuleap_oauth/checks/IDTokenChecker.java
Modified file: src/test/java/io/jenkins/plugins/tuleap_oauth/TuleapSecurityRealmTest.java
*************
```
Produced the following [PR](https://github.com/Raunak80Madan/metadata-plugin-modernizer/pull/2)

jonesbusy commented:
I'm not super fan of having the timestamp on the file name. It will require parsing to extract it

```
tuleap-oauth/modernization-metadata-2025-06-09T01-19-50
```
What if we just create a subfolder?

Like

```
tuleap-oauth/modernization-metadata/2025-06-09T01-19-50.json
```
Also perhaps the JSON extension would be useful for just display it correctly on GitHub and any IDEA?

I'm fine with the rest. (perhaps it deserve some unit test as well)

Thanks

CodexRaunak commented:
> @CodexRaunak Can you also rebase on main? Or merge back. I hope test would be stable now

Sure I will do a rebase

CodexRaunak commented:
rebased:
- Added some unit tests 
- Store the modernization metadata in a subdirectory like `tuleap-oauth/modernization-metadata/2025-06-09T01-19-50.json`
- Added `.json` extension, for syntax highlighting , but still not pretty printed.
Perhaps we need to add a method to pretty print it, I will do it later tonight, do let me know if any other changes needed.

Also currently the metadata organisation is set to my id, need to change to `jenkins-infra` later.

jonesbusy commented:
The IT also need fix

![image](https://github.com/user-attachments/assets/95193a06-1ef8-46bc-b514-37c4a512e0e3)


```java
For public static final String METADATA_ORGANISATION = "Raunak80Madan";
```

I think it's worth adding a settings/config like is done for

```java
public static final String ORGANIZATION = getTargetOrganisation();
```

CodexRaunak commented:
- Added the `pretty print json`, now we are also getting it pretty printed locally and pr content also changed - https://github.com/Raunak80Madan/metadata-plugin-modernizer/pull/2/files
- Added a method `getMetadataTargetOrganisation()` in Settings.

As in the `getTargetOrganisation()` we have a option to pass `targetOrganisation` as an `env`, else it defaults to `jenkinsci`.
```
 private static String getTargetOrganisation() {
        String targetOrganisation = System.getenv("GH_TARGET_ORGANISATION");
        if (targetOrganisation == null) {
            targetOrganisation = "jenkinsci";
        }
        return targetOrganisation;
    }
``` 
Should we have a similar option in `getMetadataTargetOrganisation()`, would the user require to set his custom repo to collect the metadata rather than using ours?

jonesbusy commented:
Sure looks good!

CodexRaunak commented:
I think it's good to go now
